drivers: gpio: add new driver for STMPE1600 gpio expander

titouanc opened the pull request:
Add support for the STMPE1600 GPIO expander (see https://www.st.com/en/interfaces-and-transceivers/stmpe1600.html).

This first version of the driver does not support interrupts because I do not have any design at hand with the interrupt pin wired to test the feature.

titouanc committed:
drivers: gpio: add new driver for STMPE1600 gpio expander

The STMPE1600 is an I2C based GPIO expander. This initial patch
only supports reading from/writing to the pins on the STMPE1600,
and there is currently no support for interrupts.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc commented:
I just fixed the issues with coding guidelines and rebased on top of master

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+	if (ret) {
+		return ret;
+	}
+	ret = write_reg16(dev, REG_GPSR_LSB, reg_state);
```
We should write to the GPSR register only if `GPIO_OUTPUT_INIT_HIGH` or `GPIO_OUTPUT_INIT_LOW` flag was set. User may choose to provide none of the two flags.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	uint16_t reg_dir = 0, reg_state = 0;
```
The values of GPSR, GPDR registers need to be cached by the driver and `reg_dir`, `reg_state` should be initialized with those cached values. Otherwise every time the function is called you're going to configure correctly `pin` but set all the remaining pins as an input. I.e. I2C transaction will reprogram full content of GPDR register, not a single bit.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	uint16_t reg_dir = 0, reg_state = 0;
+	int ret = 0;
```
We don't need to initialize `ret`, i.e. `int ret;` is enough. Here and in a few more places in this file.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+
+	k_sem_init(&drvdata->lock, 1, 1);
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+	ret = read_reg16(dev, REG_CHIP_ID_LSB, &chip_id);
+	k_sem_give(&drvdata->lock);
```
Driver initialization happens during boot time, in a single thread. There is no need to take and release the semaphore.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+	int ret = 0;
+
+	ret = read_reg16(dev, REG_GPDR_LSB, &reg_dir);
+	if (ret) {
```
Let's keep it consistent with the rest of the file and write `if (ret != 0) {`. Here and in a few more places.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+	k_sem_take(&drvdata->lock, K_FOREVER);
+
+	if (mask != 0xffff) {
+		ret = read_reg16(dev, REG_GPSR_LSB, &buf);
```
It would be better to cache the state of the register in the driver. Reading the value of the register over I2C every time is not efficient.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,297 @@
+	.port_get_raw = stmpe1600_port_get_raw,
+	.port_set_masked_raw = stmpe1600_port_set_masked_raw,
+	.port_set_bits_raw = stmpe1600_port_set_bits_raw,
+	.port_clear_bits_raw = stmpe1600_port_clear_bits_raw,
```
You need to add a dummy `pin_interrupt_configure` function that returns `-ENOTSUP`.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,288 @@
+		return ret;
+	}
+	if (set_state) {
+		ret = write_reg16(dev->config, REG_GPSR_LSB, drvdata->GPSR);
```
We need to program `GPSR` register before `GPDR` or otherwise we may observe glitch on the output pin.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,288 @@
+		return -EWOULDBLOCK;
+	}
+
+	drvdata->GPSR = (drvdata->GPSR & ~mask) | (mask & value);
```
We can update value of `drvdata->GPSR` only if we successfully written it to the device, i.e. only if call to `write_reg16` returned 0. You also need to be careful about the usage of semaphores.

mnkp reviewed (changes requested):
LGTM, few remaining issues. @titouanc could you verify that the driver passes `tests/drivers/gpio/gpio_api_1pin/` testcase?

titouanc commented:
Thank you very much for your reviews @mnkp. I just pushed the requested changes, but I still need to find some time to run the tests you suggested.

Would you need me to rebase (and possibly squash) my commits onto master ?

mnkp reviewed (commented):
A few more comments I overlooked in the original review.

> Would you need me to rebase (and possibly squash) my commits onto master ?

Yes, please rebase and squash the commits.

titouanc commented:
Thank you again for the reviews and help @mnkp.

I squashed and rebased the commits on top of `main`, including your latest remarks.

I also noticed, while running `tests/drivers/gpio/gpio_api_1pin/` that I needed to add an implementation for the API func `port_toggle_bits`. With these changes, the entire test suite is passing.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,324 @@
+			GPSR &= ~BIT(pin);
+			set_state = true;
+		}
+	}
```
It would be cleaner if we ignored `GPIO_INPUT` flag and added `else` branch here. I.e. let's remove 
```
	if ((flags & GPIO_INPUT) != 0U) {
		GPDR &= ~BIT(pin);
	}
```
and instead add here
```
	} else {
		GPDR &= ~BIT(pin);
	}
```
Since it's in fact `GPIO_OUTPUT` flag that exclusively controls the setting of the GPDR register.

mnkp commented on drivers/gpio/gpio_stmpe1600.c:
```diff
@@ -0,0 +1,324 @@
+
+static int stmpe1600_init(const struct device *dev)
+{
+	int ret;
+	uint16_t chip_id;
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
```
Looking at the source code of other gpio drivers we mostly start the variable definition block with getting pointers to `dev->config` and `dev->data` followed by remaining variable definitions and `ret` the last in the list. Let's write this block as
```
	const struct stmpe1600_config *const config = dev->config;
	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
	uint16_t chip_id;
	int ret;
```
Here and in the remaining functions. Just to keep the style aligned with other drivers.

mnkp reviewed (commented):
LGTM. Some last, minor, cosmetic tweaks.

titouanc commented:
Hello @mnkp,
I forgot to highlight you the last time I pushed my amended changes, just wanted to be sure that you see them.
Thanks again for your review !
